matgen/generate_matrix_he_and_tz.cc

```diff
@@ -0,0 +1,409 @@
+                                    1, Aij.nb(), zero, zero,
+                                    &Aij.at( ii, 0 ), Aij.stride() );
+                            }
+                        }
+                    }
+                }
+                i_global += A.tileMb( i );
+            }
+        }
+    }
+
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Generates an m-by-n Hermitian-storage test matrix.
+/// Handles Hermitian matrices.
+/// Diagonal elements of a Hermitian matrix must be real;
+/// their imaginary part must be 0.
+/// @see generate_matrix
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<scalar_t>& A,
+    std::vector< blas::real_type<scalar_t> >& Sigma,
+    slate::Options const& opts)
+{
+    slate::BaseTrapezoidMatrix<scalar_t>& TZ = A;
+    generate_matrix( params, TZ, Sigma, opts );
+
+    // Set diagonal to real.
+    #pragma omp parallel for
+    for (int64_t i = 0; i < A.mt(); ++i) {
+        if (A.tileIsLocal( i, i )) {
+            A.tileGetForWriting( i, i, LayoutConvert::ColMajor );
+            auto T = A( i, i );
+            int64_t mb = T.mb();
+            for (int64_t ii = 0; ii < mb; ++ii) {
+                T.at( ii, ii ) = std::real( T( ii, ii ) );
+            }
+        }
+    }
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Overload without Sigma.
+/// @see generate_matrix()
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<scalar_t>& A,
+    slate::Options const& opts)
+{
+    using real_t = blas::real_type<scalar_t>;
+    std::vector<real_t> dummy;
+    generate_matrix( params, A, dummy, opts );
+}
+
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::BaseTrapezoidMatrix<scalar_t>& A,
+    slate::Options const& opts)
+{
+    using real_t = blas::real_type<scalar_t>;
+    std::vector<real_t> dummy;
+    generate_matrix( params, A, dummy, opts );
+}
+
+//------------------------------------------------------------------------------
+// Explicit instantiations - hermitian matrix.
+template
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<float>& A,
+    slate::Options const& opts);
+
+template
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<double>& A,
+    slate::Options const& opts);
+
+template
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix< std::complex<float> >& A,
+    slate::Options const& opts);
+
+template
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix< std::complex<double> >& A,
+    slate::Options const& opts);
+
+//------------------------------------------------------------------------------
```

Comment:
@mgates3 I have noticed for some instances we use one of the two: 

1.
```
//------------------------------------------------------------------------------
// Documenation/Comment
```
2. 
```
//------------------------------------------------------------------------------
/// Documentation/Comment
///
```
Is there a preference when using one or the other? Just so I know moving forward.

Reply:
SLATE is documented with Doxygen.  So, all of the comments starting with `///` will be compiled into [the documentation](https://icl.bitbucket.io/slate/).  (That's why there are all lines starting with `@`; they're formatting commands for Doxygen.)  It should appear right before a the definition of a function/class/etc. and should contain information for users.  (In your above question about the word "docs", the `///` comments are the docs.)

Comments starting with just `//` are regular comments that are ignored by Doxygen.  So, it'd be things just meant for developers, like labeling our explicit instantiations as explicit instantiations.

Reply:
Thank you for the explanation @neil-lindquist.